### PR TITLE
Fix tcp disconnect on windows

### LINKED ISSUE #14: Improper stream disconnection on Windows (Hub leaks streams)
USING: Windows 10 Pro 64-bit  OS

On the file 'dashel/dashel/dashel-win32.cpp ', there is a bug on the method 'bool Hub::step(const)' which incurs in not closing appropriately clients connections, and therefore keeps listening to the past clients whose connection should had been destroyed. 

This generates either a collapse of the hub method  (since a 64 position vector gets filled up), or a timeout on the connection.

Our hypothesis is that there's  a function called 'notifyEvent' which changes the vector ets[ ], thus changing the behavior of the conditionals that decide when to connect,receive data and disconnect. We changed the position in which this function is called, an managed to make it "kind" of work. A client can indeed connect and disconnect from the hub but the first time it connects it produces a timeout error. From here on it continues working. 

This happens because when Client 1 connects at event = 0  and Client 2 connects at event = 1 and so on, each client should disconnect on the same event in which it made the connection. However Client 1 is disconnecting at event = 1 and Client 2 at event = 2; so i think this is the mistake

### PR DESCRIPTION
**Pull request by cor3ntin**
Fix #14

### CODE REVIEW

File: dashel/dashel-win32.cpp [hunk 1]
```diff
@@ -1083,7 +1086,7 @@ namespace Dashel
 			target.add(params.c_str());
 
 			sock = getOrCreateSocket(target);
-			if (target.get<SOCKET>("sock") >= 0)
+            if (target.get<SOCKET>("sock") != INVALID_SOCKET)
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: buggy indentation

File: dashel/dashel-win32.cpp [hunk 2]
```diff
@@ -1167,7 +1170,7 @@ namespace Dashel
 			if(!readyToRead)
 			{
 				// Block until something happens.
-				WaitForSingleObject(hev, INFINITE);
+				WaitForSingleObject(hev, DEFAULT_WAIT_TIMEOUT);
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: why not infinite?

File: dashel/dashel-win32.cpp [hunk 3]
```diff
@@ -1440,18 +1443,19 @@ namespace Dashel
 	{
 		while(step(-1));
 	}
-	
+
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: buggy indentation

File: dashel/dashel-win32.cpp [hunk 4]
```diff
@@ -1440,18 +1443,19 @@ namespace Dashel
...
+		lock();
+
+		std::vector<HANDLE> hEvs(streams.size());
+		std::vector<EvType> ets(streams.size());
+		std::vector<WaitableStream*> strs(streams.size());
+
 		// Wait on all our events.
 		DWORD ms = timeout >= 0 ? timeout : INFINITE;
-		
+
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: buggy indentation

File: dashel/dashel-win32.cpp [hunk 5]
```diff
@@ -1463,11 +1467,10 @@ namespace Dashel
 				WaitableStream* stream = polymorphic_downcast<WaitableStream*>(*it);
 				for(std::map<EvType,HANDLE>::iterator ei = stream->hEvents.begin(); ei != stream->hEvents.end(); ++ei)
 				{
-					// check array bounds, abort cleanly instead of creating memory trash
-					if (hc == 64)
-					{
-						std::cerr << "Trying to wait on more than 64 events" << std::endl;
-						abort();
+					if(hEvs.size() < hc) {
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: missing space between if and (

File: dashel/dashel-win32.cpp [hunk 6]
```diff
@@ -1561,6 +1564,11 @@ namespace Dashel
 						closeStream(strs[r]);
 					}
 				}
+				if(ets[r] == EvClosed)
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: missing space between if and (

File: dashel/dashel-win32.cpp [hunk 7]
```diff
@@ -1480,7 +1483,7 @@ namespace Dashel
 			unlock();
 
 			// force finite timeout to check for serial disconnections
-			DWORD r = WaitForMultipleObjects(hc, hEvs, FALSE, ms == INFINITE ? 1000 : ms);
+			DWORD r = WaitForMultipleObjects(hc, hEvs.data(), FALSE, ms == INFINITE ? DEFAULT_WAIT_TIMEOUT : ms);
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: I think that the 1000 value was added to let the Dashel Win32 hub to check periodically for device disconnection. How large is `DEFAULT_WAIT_TIMEOUT`?

File: dashel/dashel-win32.cpp (hunk 5 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: Fixed

File: dashel/dashel-win32.cpp (hunk 6 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: Fixed

File: dashel/dashel-win32.cpp (hunk 1 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: Fixed

File: dashel/dashel-win32.cpp (hunk 4 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: I was buggy before !

File: dashel/dashel-win32.cpp (hunk 7 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: 1000, i extracted it to a constant, it could be shorter though

File: dashel/dashel-win32.cpp (hunk 6 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: fixed ( weird convention, it will take some getting used to)

File: dashel/dashel-win32.cpp (hunk 2 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: hum, yeah, this one could (probably should ) be infinite.

File: dashel/dashel-win32.cpp (hunk 3 above)
Comment (cor3ntin) [reply dashel/dashel-win32.cpp]: Arguably a WS-only change, but it was buggy before

**Comment by cor3ntin** [comment]
I fix hopefully most of your remark. Well, I actually reworked the patch significantly, there was initially a number of critical bugs beside formatting issue.

Two things happened:
 * FD_CLOSE Was not listen for after a read.
 * More importantly, we detect the socket is disconnected  *during* the handling of a "potentialdata" event.
The root cause if a bug was an order of operation issue between the call to `notifyEvent` and the handling of the `EvClosed` event ( the former *must happen before the later ).

**Comment by davidjsherman** [comment]
> an order of operation issue between the call to `notifyEvent` and the handling of the `EvClosed` event
I was hoping that it would be something like that. Bravo!

File: dashel/dashel-win32.cpp (hunk 3 above)
Comment (stephanemagnenat) [reply dashel/dashel-win32.cpp]: In Dashel (and Aseba), the current style is to keep indentation on empty lines. I know that currently this style is out of fashion, but then it was common. Also, I find the out-of-fashionableness somewhat arbitrary.

File: dashel/dashel-win32.cpp (hunk 4 above)
Comment (stephanemagnenat) [reply dashel/dashel-win32.cpp]: In Dashel (and Aseba), the current style is to keep indentation on empty lines. I know that currently this style is out of fashion, but then it was common. Also, I find the out-of-fashionableness somewhat arbitrary.

File: dashel/dashel-win32.cpp (hunk 7 above)
Comment (stephanemagnenat) [reply dashel/dashel-win32.cpp]: ok, good!

**Pull request merged**